arbor/threading/threading.hpp

```diff
@@ -146,21 +174,22 @@ class task_group {
     using callable = typename std::decay<F>::type;
 
     template <typename F>
-    wrap<callable<F>> make_wrapped_function(F&& f, std::atomic<std::size_t>& c) {
-        return wrap<callable<F>>(std::forward<F>(f), c);
+    wrap<callable<F>> make_wrapped_function(F&& f, std::atomic<std::size_t>& c, exception_struct& ex) {
+        return wrap<callable<F>>(std::forward<F>(f), c, ex);
     }
 
     template<typename F>
     void run(F&& f) {
         ++in_flight_;
-        task_system_->async(make_wrapped_function(std::forward<F>(f), in_flight_));
+        task_system_->async(make_wrapped_function(std::forward<F>(f), in_flight_, exception_status_));
     }
 
     // wait till all tasks in this group are done
-    void wait() {
+    std::exception_ptr wait() {
```

Comment:
I think it makes more sense to check for an exception here, and rethrow it.
Otherwise we will have exceptions being silently ignored.
@halfflat, what are your feelings about this?

Our motivating example, where an exception thrown during model building from a recipe would still not work as expected, without adding tests to the return value of every call to `wait` inside `simulation`.

Reply:
An alternative, would be to make the destructor of `exception_struct` rethrow the exception if it has not explicitly been acquired or rethrown.
That way, if we ignore the return value of `task_group::wait()`, any exception will be rethrown when it is destructed.